### PR TITLE
Initial node performance testing framework.

### LINKED ISSUE #65249: Node Performance Testing.
# Background
1. We have some low level node performance optimization features already implemented in the Kubelet (e.g., CPU manager).
2. We don't have a way to objectively measure how well certain performance-sensitive applications run on k8s across versions.
3. We don’t have performance data on how performance-sensitive applications benefit from the already implemented features.
4. We don't have a way to measure the effect of further performance features on the classes of performance sensitive applications that our users care about. This leads to making decisions based on speculative analysis of performance impact, or inability to make decisions.
5. In March 2018 at the latest resource management working group face-to-face meeting, we agreed to prioritize building a node performance testing framework and suite over some of the performance optimization proposals. This way we can make better decisions based on objective data.

# Goals
1. Detect performance regressions.
2. Provide objective data to guide node performance optimization features.

# Proposal Document
https://docs.google.com/document/d/14SfsL1Bs5jEniNaYN7kgRan9m5ZhuyLPBPvTrZ1mK94/edit#

# Phase 1 Action/Exploration Items
1. Node e2e test driver.
2. Determine initial set of workloads.
3. Add workload container images for node performance testing.
4. Documentation.
5. Periodically run the node perf testing suite with initial set of workloads. 
6. Represent node performance test results in one of the test dashboards.
    1. Choose either https://k8s-testgrid.appspot.com/ or http://node-perf-dash.k8s.io. 
    2. Initial result will be FAILED or PASSED. 
        1. Use some simple stats (e.g., the test failed if performance of workload X falls outside of -5% performance from the average performance of the last 10 runs of the workload).
    3. Resources
        1. Node-perf-dash: https://github.com/kubernetes/contrib/tree/master/node-perf-dash
        2. Testgrid: https://github.com/kubernetes/test-infra/tree/master/testgrid
        3. Compare between two test runs: https://github.com/kubernetes/perf-tests/tree/master/compare. 
        4. How PASSED or FAILED is posted now: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/gubernator.sh#L152-L157
        5. Do we need to use NodePerfData type: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/perftype/perftype.go#L38-L45

# Pull Requests
1. Adding docker images for workloads: https://github.com/kubernetes/kubernetes/pull/65251.
2. The node performance testing framework: https://github.com/kubernetes/kubernetes/pull/65250.

/assign @balajismaniam @ConnorDoyle. 
CC @vishh @derekwaynecarr

### PR DESCRIPTION
**Pull request by balajismaniam**
This PR adds a framework for node performance testing. 
Partially fixes: https://github.com/kubernetes/kubernetes/issues/65249.
Use the following command to run this test:
```sh
make test-e2e-node FOCUS="Node Performance Testing" SKIP="" PARALLELISM=1
```
It has been tested in the following environment:
- n1-standard-16
- Ubuntu 16.04
- docker 17.03.2

Note to reviewers:
This PR won't pass node e2e since the docker images in https://github.com/kubernetes/kubernetes/pull/65251 are required for this to function. The node e2e will fail when trying to pull the required images for testing.

### CODE REVIEW

**Comment by balajismaniam** [comment]
CC @ConnorDoyle

File: test/e2e_node/node_perf_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+				framework.ExpectNoError(err)
+				// Set the Kubelet config required for this workload.
+				setKubeletConfig(f, newCfg)
+
+				By("running the workload and waiting for success")
+				// Make the pod for the workload.
+				pod := makeNodePerfPod(workload)
+
+				// Create the pod.
+				startTime := time.Now()
```
Comment (ConnorDoyle) [inline test/e2e_node/node_perf_test.go]: Note: there may be additional time between when the pod is written to the API server and when Kubelet spawns the workload process. Can we look into this and try to make the timings a bit more deterministic?

**Comment by ConnorDoyle** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/66075

**Review by RenaudWasTaken — APPROVED** [review]
/lgtm

Big fan of this very clean model!

**Comment by balajismaniam** [comment]
Thanks for the review @RenaudWasTaken. We will have to wait for https://github.com/kubernetes/kubernetes/pull/65251 before merging this PR. 

/hold
/release-note-none
/milestone v1.12

**Comment by balajismaniam** [comment]
@dims pull-kubernetes-node-e2e will fail for this PR because it is trying to pull images from https://github.com/kubernetes/kubernetes/pull/65251. Those docker images are used as workloads for node performance testing.

**Comment by balajismaniam** [comment]
Per https://github.com/kubernetes/kubernetes/pull/65251#issuecomment-417496902, the node e2es should pass now. 

/retest

**Comment by dims** [comment]
@balajismaniam yes, i requested @cjwagner to build the images :)

**Comment by dims** [comment]
@balajismaniam if you rebase this to master you can see why node-e2e is failing. please fix.

```
[dims@dims-mac 09:27] ~/go/src/k8s.io/kubernetes ⟩ go test -v ./test/e2e_node/
# k8s.io/kubernetes/test/e2e_node
test/e2e_node/node_perf_test.go:25:2: cannot find package "k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig" in any of:
	/Users/dims/go/src/k8s.io/kubernetes/vendor/k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig (vendor tree)
	/usr/local/Cellar/go/1.10.2/libexec/src/k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig (from $GOROOT)
	/Users/dims/go/src/k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig (from $GOPATH)
FAIL	k8s.io/kubernetes/test/e2e_node [setup failed]
```

/lgtm cancel

**Pull request merged**